starky/src/recursive_verifier.rs

```diff
@@ -0,0 +1,261 @@
+use itertools::Itertools;
+use plonky2::field::extension_field::Extendable;
+use plonky2::field::field_types::Field;
+use plonky2::fri::witness_util::set_fri_proof_target;
+use plonky2::hash::hash_types::RichField;
+use plonky2::iop::ext_target::ExtensionTarget;
+use plonky2::iop::witness::Witness;
+use plonky2::plonk::circuit_builder::CircuitBuilder;
+use plonky2::plonk::config::{AlgebraicHasher, GenericConfig};
+use plonky2::util::reducing::ReducingFactorTarget;
+
+use crate::config::StarkConfig;
+use crate::constraint_consumer::RecursiveConstraintConsumer;
+use crate::proof::{
+    StarkOpeningSetTarget, StarkProof, StarkProofChallengesTarget, StarkProofTarget,
+    StarkProofWithPublicInputs, StarkProofWithPublicInputsTarget,
+};
+use crate::stark::Stark;
+use crate::vars::StarkEvaluationTargets;
+
+pub fn recursively_verify_stark_proof<
```

Comment:
Random thought: Maybe we should have a naming convention for gadgets not in an `impl CircuitBuilder` block. 
Here I've used `recursively_*` to distinguish from the native stark verifier. Other options would be `circuit_*` or `gadget_*`. Can also work as suffixes.

Reply:
Yeah agreed. I think suffixes can be nice when we have native and recursive variants of something, to emphasize their similarity, e.g. they'd be listed side by side on docs.rs.

Looks like we have some existing methods with `_recursive` and some with `_recursively`, so we should probably change one to match the other, or `_circuit` seems reasonable too if you prefer it.

Separately, maybe rename the native verification function to `verify_stark_proof` to match this one?

Reply:
> Yeah agreed. I think suffixes can be nice when we have native and recursive variants of something, to emphasize their similarity, e.g. they'd be listed side by side on docs.rs.

Good idea!

> Looks like we have some existing methods with `_recursive` and some with `_recursively`, so we should probably change one to match the other, or `_circuit` seems fine as well if you prefer it.

I have a slight preference for `_circuit`, since some gadgets might not be directly used in recursion. But I'd be fine with the other suffixes too. 

> Separately, maybe rename the native verification function to `verify_stark_proof` to match this one?

Good idea, done.

I'll leave it like this for now and make another PR where I'll standardize the gadgets suffix.

Reply:
Yeah true, `_circuit` sounds good then.